### PR TITLE
Sink

### PR DESCRIPTION
**Pull request by CJPoll**
Joins + SQLSink

### CODE REVIEW

File: lib/rivulet/join/elastic_search.ex [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      %{mappings:
+        %{join_documents:
+          %{properties:
+            %{join_key: %{type: "keyword"},
+              document: %{type: "object"}}}}}
+      |> JSON.encode
+    put("/#{index_name}", json, [])
+  end
+
+  def put_join_doc(join_id, join_key, object_id, document) do
```
Comment (aweagel) [inline lib/rivulet/join/elastic_search.ex]: Consider dropping this function

File: lib/rivulet/sql_sink.ex [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    Logger.debug("Getting value schema for subject: #{subject}")
+    {:ok, %Rivulet.Avro.Schema{} = schema} = Rivulet.Avro.schema_for_subject(subject)
+
+    ensure_table_created!(config, schema.schema)
+  end
+
+  def ensure_table_created!(%Config{} = config, %AvroEx.Schema{} = schema) do
+    Logger.debug("Ensuring table is created with config: #{inspect config}")
+
+    #{:ok, results} =
```
Comment (aweagel) [inline lib/rivulet/sql_sink.ex]: Investigate these comments

File: lib/rivulet/sql_sink/database/sql_generator.ex [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        ({name, type}) ->
+          column_definition(name, :not_nullable, type, _primary_key = false)
+      end)
+
+    unique_index_definitions =
+      Enum.map(table.unique_constraints, fn(constraint) when is_list(constraint) ->
+        index_name_columns = Enum.join(constraint, "_")
+        index_columns = Enum.join(constraint, ",")
+        """
+        CREATE UNIQUE INDEX IF NOT EXISTS "#{table.name}_#{index_name_columns}_index" ON "#{table.name}"
```
Comment (aweagel) [inline lib/rivulet/sql_sink/database/sql_generator.ex]: oob:

Future things to do:
- Delegating the responsibility of creating migrations to the consumer
- Adding column white-listing functionality

File: lib/rivulet/sql_sink/database/table.ex [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  def with_column(%__MODULE__{} = table, name, type) do
+    %__MODULE__{table | column_definitions: [{name, type} | table.column_definitions]}
+  end
+
+  def primary_keys(%__MODULE__{} = table, primary_keys)
+  when is_list(primary_keys) or primary_keys == :sequence do
+    %__MODULE__{table | primary_keys: primary_keys}
+  end
+
+  def reverse_columns(%__MODULE__{} = table) do
```
Comment (aweagel) [inline lib/rivulet/sql_sink/database/table.ex]: Consider removing

File: lib/rivulet/sql_sink/writer.ex [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    decoded_messages = decoded_messages(messages, partition)
+
+    decoded_messages
+    |> upserts
+    |> do_upsert(state, partition)
+
+    decoded_messages
+    |> deletions
+    |> do_deletes(state, partition)
```
Comment (aweagel) [inline lib/rivulet/sql_sink/writer.ex]: Wrap in a transaction

File: lib/rivulet/sql_sink/writer.ex [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    Rivulet.Consumer.ack(Rivulet.client_name!, partition, offset)
```
Comment (aweagel) [inline lib/rivulet/sql_sink/writer.ex]: This exists! (see other note)

File: lib/rivulet/sql_sink/writer.ex [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  def do_deletes(deletions, %Config{} = config, %Partition{} = partition) when is_list(deletions) do
+    import Ecto.Query
+    delete = from t in table_name(config, partition), where: field(t, ^config.delete_key_field) in ^deletions
+    config.repo.delete_all(delete)
+  end
+
+  def do_upsert(upserts, %Config{} = config, %Partition{} = partition) do
+    config
+    |> table_name(partition)
+    |> config.repo.insert_all(upserts, on_conflict: :replace_all, conflict_target: List.flatten(config.unique_constraints))
```
Comment (aweagel) [inline lib/rivulet/sql_sink/writer.ex]: Not a blocker, but we should talk to Brian about potential alternatives for `replace_all`

File: lib/rivulet/sql_sink/consumer.ex [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+          offset_commit_policy: :commit_to_kafka_v2,
+          offset_commit_interval_seconds: 1,
+        ],
+        consumer_config: [begin_offset: :earliest, max_bytes: 2_000_000]
+      }
+
+    Consumer.start_link(__MODULE__, consumer_config, {config.table_pattern, sink})
+  end
+
+  def set_pool(consumer, pool) do
```
Comment (aweagel) [inline lib/rivulet/sql_sink/consumer.ex]: Drop this function

File: lib/rivulet/sql_sink/database/table.ex (hunk 4 above)
Comment (CJPoll) [reply lib/rivulet/sql_sink/database/table.ex]: Taking a second look, I think we should keep this function in our API, even if we're currently not using it.

File: lib/rivulet/sql_sink/writer.ex (hunk 5 above)
Comment (CJPoll) [reply lib/rivulet/sql_sink/writer.ex]: Done

File: lib/rivulet/join/funcs.ex [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    bulk_doc = ElasticSearch.bulk_put_join_doc(bulk, join_id)
+
+    Rivulet.Join.Batcher.batch_commands(batcher, bulk_doc, join_keys, topic, partition, offset)
+  end
+
+  @type ignored :: term
+
+  @spec transforms([[term]], {module, [{Rivulet.Kafka.Partition.topic, :key | :random}]})
+  :: ignored
+  def transforms(join_docs, transformers) do
```
[the remark this answers is not in the document]
Comment (CJPoll) [reply lib/rivulet/join/funcs.ex]: I think this is ok to leave here for now.

**Pull request merged**